File: javascript/binary_tree.js
const Maze = require("./Maze");
const Direction = require('./direction');

// Binary Tree algorithm is one of the simplest maze generation algorithms:
//
// 1. start at a corner of the maze (in this case it will be the North West)
// 2. iterate through the cells row by row
// 3.     for each cell pick a random East or South wall to remove
// 4.     repeat until all cells have been visited

/* returns a random integer between 0 (inclusive) and max(exclusive) */
function rand(max) {
    return Math.floor(Math.random() * max );
}


function generate(width, height) {
    const east_most_cell = (x) => x === width - 1;
    const south_most_cell = (y) => y === height - 1;

    let maze = new Maze(width, height);
    const es_directions = [Direction.E, Direction.S];

    for (let y = 0; y < maze.height; y++) {
        for (let x = 0; x < maze.width; x++) {
            if (east_most_cell(x) && south_most_cell(y)) {
                continue;
            } else if (east_most_cell(x)) {
                maze.field[y][x] |= Direction.S;
            } else if (south_most_cell(y)) {
                maze.field[y][x] |= Direction.E;
            } else {
                maze.field[y][x] |= es_directions[rand(2)];
            }
        }
    }
    return maze;
}

width = process.argv.length > 2 ? parseInt(process.argv[2], 10) : 20
height = process.argv.length > 3 ? parseInt(process.argv[3], 10) : 20
maze = generate(width, height);
maze.display();

File: python/binary_tree.py
from maze import (Maze, Direction)
import random
import sys

# Binary Tree algorithm is one of the simplest maze generation algorithms:
#
# 1. start at a corner of the maze (in this case it will be the North West)
# 2. iterate through the cells row by row
# 3.     for each cell pick a random East or South wall to remove
# 4.     repeat until all cells have been visited


def generate(width, height) -> Maze:

    def east_most_cell(x):
        return x == width - 1

    def south_most_cell(y):
        return y == height - 1

    maze = Maze(width, height)

    es_directions = [Direction.E, Direction.S]

    for y in range(height):
        for x in range(width):
            if east_most_cell(x) and south_most_cell(y):
                continue
            elif east_most_cell(x):
                maze.field[y][x] |= Direction.S.value
            elif south_most_cell(y):
                maze.field[y][x] |= Direction.E.value
            else:
                random_dir = random.choice(es_directions)
                maze.field[y][x] |= random_dir.value

    return maze


width = int(sys.argv[1]) if len(sys.argv) > 1 else 20
height = int(sys.argv[2]) if len(sys.argv) > 2 else 20
maze = generate(width, height)
maze.display()
